src/krb5_plugin/idp/idp_utils.c

```diff
@@ -0,0 +1,387 @@
+
+    /* These are required fields. */
+    if (is_empty(verification_uri) || is_empty(user_code)) {
+        return NULL;
+    }
+
+    data = malloc(sizeof(struct sss_idp_oauth2));
+    if (data == NULL) {
+        return NULL;
+    }
+    memset(data, 0, sizeof(struct sss_idp_oauth2));
+
+    data->verification_uri = strdup(verification_uri);
+    data->user_code = strdup(user_code);
+    if (data->verification_uri == NULL || data->user_code == NULL) {
+        sss_idp_oauth2_free(data);
+        return NULL;
+    }
+
+    if (!is_empty(verification_uri_complete)) {
+        data->verification_uri_complete = strdup(verification_uri_complete);
+        if (data->verification_uri_complete == NULL) {
+            sss_idp_oauth2_free(data);
+            return NULL;
+        }
+    }
+
+    return data;
+}
+
+static struct sss_idp_oauth2 *
+sss_idp_oauth2_from_json(const char *json_str)
+{
+    struct sss_idp_oauth2 jdata = {0};
+    struct sss_idp_oauth2 *data;
+    json_error_t jret;
+    json_t *jroot;
+    int ret;
+
+    jroot = json_loads(json_str, 0, &jret);
+    if (jroot == NULL) {
+        return NULL;
+    }
+
+    ret = json_unpack(jroot, "{s:s, s?:s, s:s}",
+                "verification_uri", &jdata.verification_uri,
+                "verification_uri_complete", &jdata.verification_uri_complete,
+                "user_code", &jdata.user_code);
+    if (ret != 0) {
+        json_decref(jroot);
+        return NULL;
+    }
+
+    data = sss_idp_oauth2_init(jdata.verification_uri,
+                               jdata.verification_uri_complete,
+                               jdata.user_code);
+
+    json_decref(jroot);
+    return data;
+}
+
+static char *
+sss_idp_oauth2_to_json(const struct sss_idp_oauth2 *data)
+{
+    json_t *jroot;
+    char *str;
+
+    if (data == NULL) {
+        return NULL;
+    }
+
+    /* These are required fields. */
+    if (data->verification_uri == NULL || data->user_code == NULL) {
+        return NULL;
+    }
+
+    jroot = json_pack("{s:s?, s:s*, s:s?}",
+                "verification_uri", data->verification_uri,
+                "verification_uri_complete", data->verification_uri_complete,
+                "user_code", data->user_code);
+    if (jroot == NULL) {
+        return NULL;
+    }
+
+    str = json_dumps(jroot, JSON_COMPACT);
+    json_decref(jroot);
+
+    return str;
+}
+
+static struct sss_idp_oauth2 *
+sss_idp_oauth2_decode(const char *str)
+{
+    size_t prefix_len;
+
+    prefix_len = strlen(SSSD_IDP_OAUTH2_PREFIX);
+    if (strncmp(str, SSSD_IDP_OAUTH2_PREFIX, prefix_len) != 0) {
+        return NULL;
+    }
+
+    return sss_idp_oauth2_from_json(str + prefix_len);
+}
+
+static char *
+sss_idp_oauth2_encode(struct sss_idp_oauth2 *data)
+{
+    char *json_str;
+    char *str;
+    int aret;
+
+    json_str = sss_idp_oauth2_to_json(data);
+    if (json_str == NULL) {
+        return NULL;
+    }
+
+    aret = asprintf(&str, "%s%s", SSSD_IDP_OAUTH2_PREFIX, json_str);
```

Comment:
Hi,

you should free(json_str) here, otherwise it is leaking and is causing the valgrind errors in the CI.

bye,
Sumit